File: stryker4jvm-core/src/main/java/stryker4jvm/core/model/MutantMetaData.java
package stryker4jvm.core.model;

import stryker4jvm.core.model.elements.Location;

/** Data class that describes the original and mutated code with strings */
public final class MutantMetaData {

  public final String original;
  public final String replacement;
  public final String mutatorName;
  public final Location location;

  public MutantMetaData(
      String original, String replacement, String mutatorName, Location location) {
    this.original = original;
    this.replacement = replacement;
    this.mutatorName = mutatorName;
    this.location = location;
  }

  public String showLocation() {
    return String.format(
        "%d:%d to %d:%d",
        location.start.line, location.start.column, location.end.line, location.end.column);
  }
}

File: modules/mutatorApi/src/main/scala/stryker4s/mutatorapi/MutantMetadata.scala
package stryker4s.mutatorapi

import mutationtesting.{Location, Position as MTPosition}

import scala.meta.Position

/** Metadata of a [[stryker4s.mutatorapi.MutatedCode]]
  *
  * @param original
  *   Original code
  * @param replacement
  *   Mutated replaced code
  * @param mutatorName
  *   Mutator category
  * @param location
  *   The location of the mutated code
  */
final case class MutantMetadata(
    original: String,
    replacement: String,
    mutatorName: String,
    location: Location,
    description: Option[String]
)

object MutantMetadata {

  /** Convenience constructor that converts a scalameta `Position` to the `Location` used in reports.
    */
  def apply(
      original: String,
      replacement: String,
      mutatorName: String,
      position: Position,
      description: Option[String]
  ): MutantMetadata =
    MutantMetadata(original, replacement, mutatorName, toLocation(position), description)

  private def toLocation(pos: Position): Location = Location(
    start = MTPosition(line = pos.startLine + 1, column = pos.startColumn + 1),
    end = MTPosition(line = pos.endLine + 1, column = pos.endColumn + 1)
  )
}
